File: src/HeavyIndustry/world/blocks/distribution/AdaptDirectionalUnloader.java
package HeavyIndustry.world.blocks.distribution;

import mindustry.world.blocks.distribution.DirectionalUnloader;

/**
 * The ultimate dream of Saturated Firepower.
 */
public class AdaptDirectionalUnloader extends DirectionalUnloader {
    public AdaptDirectionalUnloader(String name) {
        super(name);
        buildType = AdaptDirectionalUnloaderBuild::new;
    }

    public class AdaptDirectionalUnloaderBuild extends DirectionalUnloaderBuild{
        private float counter;

        @Override
        public void updateTile(){
            counter += edelta();
            final float limit = speed;

            while(counter >= limit){
                unloadTimer = speed;
                super.updateTile();
                counter -= limit;
            }
        }
    }
}

File: src/heavyindustry/world/blocks/distribution/AdaptDirectionalUnloader.kt
package heavyindustry.world.blocks.distribution

import mindustry.world.blocks.distribution.DirectionalUnloader

/**
 * A directional unloader that is not affected by game frame rates.
 * @author Wisadell
 */
open class AdaptDirectionalUnloader(name: String) : DirectionalUnloader(name) {
    open inner class AdaptDirectionalUnloaderBuild : DirectionalUnloaderBuild() {
        private var counter = 0f

        /** Make its uninstallation speed no longer affected by frame rate. */
        override fun updateTile() {
            counter += edelta()
            val limit = speed

            while (counter >= limit) {
                unloadTimer = speed
                super.updateTile()
                counter -= limit
            }
        }
    }
}
